Expose VO performance data (render/present times) to users

haasn opened the pull request:
Currently exposed as a property, backed by a VOCTRL. Not sure if this is the best design, but it's the best I could come up with.

This should allow projects like https://github.com/Argon-/mpv-stats to present the user with meaningful performance metrics, similar to madVR / MPDN etc, hopefully helping users understand performance impacts of different settings and pick the right options for their setups.

haasn commented:
Oh, also, commit 71d7479 is somewhat optional. I don't know if that's a better design or not.

haasn commented:
Pretty handy command for testing:

```
T show-text "last: ${upload-time-last} ${render-time-last} ${present-time-last}\navg: ${upload-time-avg} ${render-time-avg} ${present-time-avg}\npeak: ${upload-time-peak} ${render-time-peak} ${present-time-peak}"
```

This lets me evaluate performance metrics in real-time pretty easily.

wm4 commented on video/out/opengl/utils.c:
```diff
@@ -1019,3 +1019,77 @@ void gl_sc_gen_shader_and_reset(struct gl_shader_cache *sc)
+{
+    GL *gl = timer->gl;
+    if (gl->DeleteQueries)
+        gl->DeleteQueries(QUERY_OBJECT_NUM, timer->query);
```
uninit should probably be idempotent and also work on an not initialized struct (i.e. check for gl and reset the query IDs).

haasn replied:
DeleteQueries is a no-op on unitialized queries (ID 0).

wm4 replied:
But you don't reset the IDs.

wm4 commented on video/out/opengl/common.c:
```diff
@@ -274,6 +274,23 @@ static const struct gl_functions gl_functions[] = {
+            DEF_FN(GetQueryObjectiv),
+            DEF_FN(GetQueryObjecti64v),
+            DEF_FN(GetQueryObjectuiv),
+            DEF_FN(GetQueryObjectui64v),
```
The 64 bit variants are apparently not available in GL 3.2, only 3.3.

haasn replied:
The compatibility matrix on opengl.org says 3.2, the description says 3.3. I don't know which one is accurate. Maybe we should just set it to 3.3 to be sure?

(Note: I copied these almost verbatim from one of your commits, which also used 3.2 here)

wm4 replied:
>The compatibility matrix on opengl.org says 3.2, the description says 3.3.

What is that?


> I don't know which one is accurate. Maybe we should just set it to 3.3 to be sure?

When in doubt, check the specs. They're authoritative.

haasn replied:
> What is that?

This one: https://www.opengl.org/sdk/docs/man/html/glGetQueryObject.xhtml

> When in doubt, check the specs. They're authoritative.

Okay. Spec says it's not available in 3.2

wm4 replied:
>This one: https://www.opengl.org/sdk/docs/man/html/glGetQueryObject.xhtml

That's just plain wrong, even on the same page it says:

> glGetQueryObjecti64v and glGetQueryObjectui64v are available only if the GL version is 3.3 or greater.

wm4 commented on video/out/opengl/utils.c:
```diff
@@ -1019,3 +1019,77 @@ void gl_sc_gen_shader_and_reset(struct gl_shader_cache *sc)
+
+    // Update internal sum
+    timer->avg_sum = timer->avg_sum + new - old;
+    timer->sample_count = MPMIN(timer->sample_count+1, QUERY_SAMPLE_SIZE);
```
Missing spaces around +.

wm4 commented on video/out/opengl/utils.c:
```diff
@@ -1019,3 +1019,77 @@ void gl_sc_gen_shader_and_reset(struct gl_shader_cache *sc)
+        GLuint64 elapsed;
+        gl->GetQueryObjectui64v(id, GL_QUERY_RESULT, &elapsed);
+        gl_timer_record(timer, elapsed);
+    }
```
Doesn't that create quite a delay?

haasn replied:
I'm not sure, why would it? Shouldn't GetQueryObjectui64v be essentially free on already-completed queries?

wm4 replied:
I mean, doesn't this retrieve a query only after QUERY_OBJECT_NUM queries have been used?

haasn replied:
Oh, right. Could add a loop to check for all query objects which have results available already but I don't see much of a point in carrying the complexity. Even at 8 frames delay you're only talking about like one third of a second for 24 fps content.

wm4 commented on player/command.c:
```diff
@@ -2381,6 +2381,34 @@ static int panscan_property_helper(void *ctx, struct m_property *prop,
+
+    struct voctrl_performance_data data = {0};
+    if (vo_control(mpctx->video_out, VOCTRL_PERFORMANCE_DATA, &data) <= 0)
+        return M_PROPERTY_UNAVAILABLE;
```
Keep in mind that sending voctrls every frame also have performance implications. It forces the main thread to sync with the VO thread, while the code carefully tried to avoid such blocking before. Also, since most time the VO thread is blocked on the GL swapbuffer call (for vsync), the voctrl could be significantly delayed each frame, up to the point where the main thread blocks until the VO has no frame to display anymore, unblocking the main thread for a while.

But that's not my problem fortunately.

haasn replied:
Would it be a better design to have a global struct inside MPContext or something that the VO updates regularly, with the property code only reading these fields?

wm4 replied:
> Would it be a better design to have a global struct inside MPContext

Probably not. I don't see any nice way here.

wm4 commented:
What's with this timer unit chaos? Internally, mpv tries to stick with either uint64_t microseconds, or double seconds (most of the time), and this introduces nanoseconds and milliseconds.

haasn commented:
Nanoseconds is based on what the OpenGL API itself gives me. I could change it to microseconds, since it's not like the extra precision is necessary. I just figured it wouldn't hurt.

I chose milliseconds for the properties because it's the right timescale. Performance data is usually only interesting in milliseconds, and if I set it to seconds I would have to then write a lua function to query the properties and multiply them all by 1000 before displaying. This way I can just write e.g. ```show-text "${render-time-avg}"``` and get a useful display out it.

haasn commented:
I addressed all of the mentioned suggestions, and also refactored the ugly property macro spam somewhat. Let me know if I should squash some of the commits together (not much point in including "dead" code in master's timeline)

wm4 commented:
Yes, might be better to squash them all together, maybe keeping the vo_opengl and property changes separate.

Regarding avoding voctrl, the only idea I have is making an opque struct, returning a pointer to it via voctrl, and then cache the voctrl. The struct would need accessors, which are somehow thread-safe.

haasn commented:
Regarding voctrl, I honestly don't think it's a huge issue. Even holding down a button (that internally runs this voctrl a total of 9 times) on every single frame I don't notice any change in performance. (Same when I hold down the stats.lua script)

But maybe it's more of an issue on slower CPUs?

wm4 commented:
I wouldn't worry about that. I mean, I do know that the popular lua-stats script (or whatever it was called) can severely degrade performance, but that's the user's problem.

What would be worse if the playloop is subtly prevented from decoding frames ahead of time or so.

haasn commented:
Updated. Squashed the commits and also fixed a floating point crash (+ an accidentally field in MPContext left over from a dead code path).

wm4 commented:
Merged.
